### PR TITLE
[SPARK-45536][BUILD] Lower the default `-Xmx` of `build/mvn` to 3g

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr lower the default `-Xmx` of `build/mvn` from 4g to 3g to reduce the peak memory usage of Maven compilation.

### Why are the changes needed?
This can potentially fix the snapshot build being failed: https://github.com/apache/spark/actions/runs/6502277099

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

- Manual check.

run 
```
build/mvn clean install -DskipTests -Pyarn -Pkubernetes -Pvolcano -Phive -Phive-thriftserver -Pspark-ganglia-lgpl -Phadoop-cloud
```

**Before**

Peak memory usage is at 6.1GB.

**After**

Peak memory usage is at 5GB, but the compilation time has increased by 10%.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by LuciferYang** [comment]
@HyukjinKwon Is there a way to manually trigger `Publish snapshot` using this PR, or do we have to wait until it's merged to get validation?

**Comment by HyukjinKwon** [comment]
We can merge and try. I made another PR https://github.com/apache/spark/pull/43365 for a different apprach.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by EnricoMi** [comment]
> @HyukjinKwon Is there a way to manually trigger `Publish snapshot` using this PR, ...

You could if you'd add this to the `publish_snapshot.yml` workflow before:

    on:
      workflow_dispatch

https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow

**Comment by beliefer** [comment]
@LuciferYang I don't understand `This can potentially fix the snapshot build being failed: https://github.com/apache/spark/actions/runs/6502277099`. Could you give an explanation why you reduce the `-Xmx`?

**Comment by EnricoMi** [comment]
The virtual machine building the releases has 7GB memory. The build process uses 6.1GB memory. They suspect the build process is killed because it uses too much memory.

**Comment by LuciferYang** [comment]
> @LuciferYang I don't understand `This can potentially fix the snapshot build being failed: https://github.com/apache/spark/actions/runs/6502277099`. Could you give an explanation why you reduce the `-Xmx`?

It's just a guess, based on historical experience, the compilation container being killed might be due to memory overuse, but I indeed don't have concrete evidence for this case. Do you have any better suggestions? @beliefer

**Comment by HyukjinKwon** [comment]
Let's just try. If it doesn't work we can revert

**Comment by LuciferYang** [comment]
```
[info] *** 1 TEST FAILED ***
[error] Failed tests:
[error] 	org.apache.spark.sql.kafka010.KafkaSourceStressSuite
[error] (sql-kafka-0-10 / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 2132 s (35:32), completed Oct 13, 2023, 12:07:56 PM
```

Only `KafkaSourceStressSuite` test failed, this is a known flaky test

**Comment by LuciferYang** [comment]
Merge into master to observe the `Publish Snapshot` job, if it doesn't work, we can revert it tomorrow.

Thanks @HyukjinKwon @beliefer @EnricoMi

**Comment by dongjoon-hyun** [comment]
Thank you, @LuciferYang and all.

Since Java 17 JVM GC is different than the old ParallelGC, we can optimize further.

**Comment by LuciferYang** [comment]
@dongjoon-hyun It seems to still not work. Do you have any ideas or suggestions for optimizing the compilation options?

**Comment by LuciferYang** [comment]
I tried to perform mvn deploy operation to the local nexus, and unusual happened ...

**Comment by EnricoMi** [comment]
Another attempt to fix this in #43512 / SPARK-45651.

**Pull request closed without merging**
